src/Simulation/Common/StackTrace.cs

```diff
@@ -135,36 +136,94 @@ public static StackFrame[] PopulateSourceLocations(Stack<StackFrame> qsharpCallS
-            for (int csharpStackFrameId = 0; csharpStackFrameId < csharpCallStack.Length; ++csharpStackFrameId)
+            int currentQSharpFrameIndex = 0;
+            int currentCSharpFrameIndex = 0;
+
+            System.Diagnostics.StackFrame csharpFrame = GetNextCSharpStackFrame(csharpStackFrames, ref currentCSharpFrameIndex);
+            StackFrame qsharpFrame = GetNextQSharpStackFrame(qsharpStackFrames, ref currentQSharpFrameIndex);
+            while (csharpFrame != null && qsharpFrame != null)
             {
-                string fileName = csharpCallStack[csharpStackFrameId].GetFileName();
-                if (fileName != null)
+                if (IsMatch(csharpFrame, qsharpFrame))
                 {
-                    fileName = System.IO.Path.GetFullPath(fileName);
-                    int failedLineNumber = csharpCallStack[csharpStackFrameId].GetFileLineNumber();
-                    StackFrame currentQsharpStackFrame = qsharpStackFrames[qsharpStackFrameId];
-                    if (fileName == currentQsharpStackFrame.SourceFile &&
-                        currentQsharpStackFrame.DeclarationStartLineNumber <= failedLineNumber &&
-                            (
-                                (failedLineNumber < currentQsharpStackFrame.DeclarationEndLineNumber) ||
-                                (currentQsharpStackFrame.DeclarationEndLineNumber == -1)
-                            )
-                        )
-                    {
-                        currentQsharpStackFrame.FailedLineNumber = failedLineNumber;
-                        qsharpStackFrameId++;
-                        if (qsharpStackFrameId == qsharpStackFrames.Length) break;
-                    }
+                    PopulateQSharpFrameFromCSharpFrame(csharpFrame, qsharpFrame);
+                    // Advance to the next Q# Stack frame only when it matches C# stack frame.
+                    // This assumes all Q# stack frames with enough information to match will match with corresponding C# stack frames.
+                    // If corresponding C# stack frame doesn't have enough information to match
+                    // this and all subsequent Q# stack frames will not match.
+                    qsharpFrame = GetNextQSharpStackFrame(qsharpStackFrames, ref currentQSharpFrameIndex);
                 }
+                csharpFrame = GetNextCSharpStackFrame(csharpStackFrames, ref currentCSharpFrameIndex);
             }
             return qsharpStackFrames;
         }
+
+        /// <summary>
+        /// Return next Q# stack frame that has enough information to match
+        /// </summary>
+        private static StackFrame GetNextQSharpStackFrame(StackFrame[] qsharpStackFrames, ref int currentFrameIndex)
+        {
+            while (currentFrameIndex < qsharpStackFrames.Length)
+            {
+                StackFrame currentFrame = qsharpStackFrames[currentFrameIndex];
+                currentFrameIndex++;
```

Comment:
>currentFrameIndex [](start = 16, length = 17)

currentFrame does not correspond to currentFrameIndex. It would be much nicer if they matched.

Reply:
Updated "current frame" indexes to point to current frame (and not to current+1 frame as before).